tests/cbtest/gtest/test_uzfs.cc

```diff
@@ -816,28 +1034,66 @@ TEST(uZFS, TestRebuild) {
+	execute_rebuild_test_case("rebuild error state", 2, ZVOL_REBUILDING_ERRORED, B_FALSE, B_FALSE);
 
 	/* thread helping rebuild will exit after reading REBUILD_STEP */
-	execute_rebuild_test_case("rebuild exit after step", 3, ZVOL_REBUILDING_IN_PROGRESS);
+	execute_rebuild_test_case("rebuild exit after step", 3, ZVOL_REBUILDING_IN_PROGRESS, B_FALSE, B_FALSE);
 
 	/* thread helping rebuild will exit after writng invalid write IO */
-	execute_rebuild_test_case("rebuild exit after invalid write", 4, ZVOL_REBUILDING_IN_PROGRESS);
+	execute_rebuild_test_case("rebuild exit after invalid write", 4, ZVOL_REBUILDING_IN_PROGRESS, B_FALSE, B_FALSE);
 
 	/* thread helping rebuild will exit after writng valid write IO */
-	execute_rebuild_test_case("rebuild exit after valid write", 5, ZVOL_REBUILDING_IN_PROGRESS);
+	execute_rebuild_test_case("rebuild exit after valid write", 5, ZVOL_REBUILDING_IN_PROGRESS, B_FALSE, B_FALSE);
 
 	/* thread helping rebuild will exit after writng valid write IO and REBUILD_STEP_DONE */
-	execute_rebuild_test_case("rebuild exit after valid write and rebuild_step", 6, ZVOL_REBUILDING_IN_PROGRESS);
+	execute_rebuild_test_case("rebuild exit after valid write and rebuild_step", 6, ZVOL_REBUILDING_IN_PROGRESS, B_FALSE, B_FALSE);
 
 	/* thread helping rebuild will exit after writing valid write IO and REBUILD_STEP_DONE, and reads REBUILD_STEP, writes REBUILD_STEP_DONE */
-	execute_rebuild_test_case("complete rebuild", 7, ZVOL_REBUILDING_IN_PROGRESS);
+	execute_rebuild_test_case("complete rebuild", 7, ZVOL_REBUILDING_IN_PROGRESS, B_TRUE, B_FALSE);
+	EXPECT_EQ(ZVOL_STATUS_HEALTHY, uzfs_zvol_get_status(zinfo->zv));
+
+	memset(&zinfo->zv->rebuild_info, 0, sizeof (zvol_rebuild_info_t));
+}
+
+TEST(uZFS, TestRebuildScanner) {
+	uzfs_mgmt_conn_t *conn;
+	mgmt_ack_t *mack;
+	char ip[MAX_IP_LEN];
+	kthread_t *thrd;
+	rebuild_scanner = &uzfs_zvol_rebuild_scanner;
+	rebuild_thread_arg_t *rebuild_args;
+
+	zvol_rebuild_step_size = (1024ULL * 1024ULL * 1024ULL) / 2 + 1000;
+
+	/* Rebuild thread exits abruptly just after connect */
+	execute_rebuild_test_case("Rebuild abrupt", 1,
+	    ZVOL_REBUILDING_IN_PROGRESS, B_FALSE, B_TRUE);
+
+	/* Rebuild thread sending wrong opcode after connectg */
+	execute_rebuild_test_case("Wrong opcode", 2,
+	    ZVOL_REBUILDING_IN_PROGRESS, B_FALSE, B_TRUE);
+
+	/* Rebuild thread exits after handshake */
+	execute_rebuild_test_case("Rebuild error out", 3,
+	    ZVOL_REBUILDING_ERRORED, B_FALSE, B_TRUE);
+
+	/* Rebuild thread sending wrong vol name */
+	execute_rebuild_test_case("Wrong vol name", 4,
+	    ZVOL_REBUILDING_IN_PROGRESS, B_FALSE, B_TRUE);
+
+	/* Rebuild thread sending handshake again on same volume */
+	execute_rebuild_test_case("Send handshake again", 5,
+	    ZVOL_REBUILDING_IN_PROGRESS, B_FALSE, B_TRUE);
+
+	/* Rebuild thread sendinc complete opcode */
+	execute_rebuild_test_case("complete rebuild", 6,
+	    ZVOL_REBUILDING_IN_PROGRESS, B_TRUE, B_TRUE);
```

Comment:
this approach looks wrong to me. There is a giant test which executes multiple small "tests" inside. It does not look like unit testing. Unit test should ideally test one thing. Is it possible to break it into smaller tests which are independent on each other without duplicating much code?